flink-core/src/main/java/org/apache/flink/api/common/state/MapState.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.api.common.state;
+
+import org.apache.flink.annotation.PublicEvolving;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * {@link State} interface for partitioned key-value state. The key-value pair can be
+ * added, updated and retrieved.
+ *
+ * <p>The state is accessed and modified by user functions, and checkpointed consistently
+ * by the system as part of the distributed snapshots.
+ *
+ * <p>The state is only accessible by functions applied on a KeyedDataStream. The key is
+ * automatically supplied by the system, so the function always sees the value mapped to the
+ * key of the current element. That way, the system can handle stream and state partitioning
+ * consistently together.
+ *
+ * @param <UK> Type of the keys in the state.
+ * @param <UV> Type of the values in the state.
+ */
+@PublicEvolving
+public interface MapState<UK, UV> extends AppendingState<Map<UK, UV>, Iterable<Map.Entry<UK, UV>>> {
```

Comment:
Exactly, the semantics are different from how they are described on `AppendingState` and from how it is used. `MapState` can still have a `add(Map<UK, UV> map)` method but I don't think it makes sense that it is an `AppendingState`. What are the use cases where a user is expected to have an `AppendingState<Map<K, V>, Map<K, V>>` instead of simply having a `MapState<K, V>`?

As I said before, the semantics of `MapState.add()` would be like `ValueState.update()` and `ValueState` is also not an `AppendingState`. I think a `MultiMapState` that would be like a `Map<K, Iterable<V>>` (and would have the semantics of a `ListState`) would be able to satisfy `AppendingState` but I'm not sure whether we would need the interface there.